Monitor app launch performance for iOS and Android in RUM (#33710)
DataDog/documentation: content/en/real_user_monitoring/application_monitoring/android/application_launch_monitoring.md

---
title: Android Mobile App Launch Monitoring
description: "Measure Android mobile application launch performance, including the time to initial display and time to full display."
aliases:
- /real_user_monitoring/android/application_launch_monitoring/
- /real_user_monitoring/mobile_and_tv_monitoring/application_launch_monitoring/android
- /real_user_monitoring/mobile_and_tv_monitoring/android/application_launch_monitoring
further_reading:
- link: https://github.com/DataDog/dd-sdk-android
  tag: "Source Code"
  text: Source code for dd-sdk-android
- link: /real_user_monitoring
  tag: Documentation
  text: Explore Datadog RUM
---

<div class="alert alert-danger">

>   The <code>application_start</code> action is no longer collected starting from Android SDK version 3.5.0. The <code>rum.measure.app.startup_time metric</code> is marked as deprecated but continues to report data from devices running app versions that use older SDK versions.

Reviewer: ```suggestion
  The <code>application_start</code> action is not collected in Android SDK versions 3.5.0+. The <code>rum.measure.app.startup_time metric</code> is marked as deprecated but continues to report data from devices running app versions that use older SDK versions.
```

</div>

## Overview

Application launch monitoring helps you understand how fast your Android app becomes usable after a user taps the app icon. Use it to identify slow startup times, track performance regressions, and optimize the user’s first impression of your app.

With this feature, you can:
- Measure time to initial display (TTID) and time to full display (TTFD) for cold and warm starts
- View launch performance trends in the RUM Summary and Mobile Performance Dashboard
- Drill into individual launch events to diagnose bottlenecks

## How it works

During initialization, the RUM Android SDK creates a view called `ApplicationLaunch`. This view’s start time matches the start of the Android process. The `ApplicationLaunch` view includes any logs, actions, and resources created before your first call to `startView`. 

### Time to initial display and time to full display

> Starting from Android SDK version 3.5.0, the time to initial display (TTID) and time to full display (TTFD) will be collected during the application launch period.

Reviewer: ```suggestion
In Android SDK versions 3.5.0+, the time to initial display (TTID) and time to full display (TTFD) are collected during the application launch period.
```

### Measuring the time to initial display
The Android SDK automatically measures the TTID. The TTID can be optionally profiled using the Android Mobile Profiler.

### Measuring the time to full display
The time to full display is manually defined using the `GlobalRumMonitor.get().reportAppFullyDisplayed` API in the Android SDK based on the application’s specific definition of “fully drawn.” 

Below is an example where time to full display is determined when home activity is fully loaded.

{{< tabs >}}
{{% tab "Kotlin" %}}

```kotlin
class SampleApplication : Application() {
    override fun onCreate() {
        super.onCreate()
        val configuration = Configuration.Builder(
            clientToken = "<CLIENT_TOKEN>",
            env = "<ENV_NAME>",
            variant = "<APP_VARIANT_NAME>"
        ).build()

>         Datadog.initialize(this, configuration, trackingConsent)

Author: @0xnm Yes, it should. I generated this with AI, so please let me know if it an incorrect equivalent:

```
class HomeActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_home)
        lifecycleScope.launch { loadData() }
    }

    private suspend fun loadData() {
        val url = URL("https://api.example.com/data")
        try {
            val data = withContext(Dispatchers.IO) {
                url.openConnection().getInputStream().bufferedReader().readText()
            }
            updateUI(data)
        } catch (e: Exception) {
            // Handle the error if needed
        }

        // Mark the app as fully displayed
        GlobalRumMonitor.get().reportAppFullyDisplayed()
    }
}
```

>         Datadog.initialize(this, configuration, trackingConsent)

Reviewer 2: looks good! sorry for missing this comment

>         Datadog.initialize(this, configuration, trackingConsent)

Reviewer 3: The code snippet above can work, but there are two issues as a demo in the documentation from what I see:
* It doesn't follow Google recommended MVVM pattern, since it is loading data from the View side.
* Due to the point above, we are exposing the unnecessary part (how the data is loaded) to the reader, which may confuse them.

So I suggest to change to following code snippet, `HomeViewModel` is responsible for preparing the `uiState` and the `HomeActivity` just observes the state without exposing the data loading detail, and when the state is ready, 
`reportAppFullyDisplayed` should be called.
```
class HomeActivity : AppCompatActivity() {

    private val viewModel: HomeViewModel by viewModels()

    override fun onCreate(savedInstanceState: Bundle?) {
        setContentView(R.layout.activity_home)
        super.onCreate(savedInstanceState)

        viewModel.uiState.observe(this) { state ->
            if (state.isLoaded) {
                 GlobalRumMonitor.get().reportAppFullyDisplayed()
            }
        }
    }
}
```

    }
}
```
{{% /tab %}}
{{< /tabs >}}

> If you use reportFullyDrawn to identify the moment of full display, you can use getFullyDrawnReporter to subscribe to `reportFullyDrawn` and call `RumMonitor.reportFullyDrawn`. If you do not use reportFullyDrawn, you can still call `RumMonitor.reportFullyDrawn`, but it will make sense only for cold and warm starts, rather than hot starts.

Reviewer 2: This is not the actual API. Where `getFullyDrawnReporter` is coming from?

<div class="alert alert-danger">
  If the time to full display is not defined, the Android SDK only collects the TTID.
</div>

### RUM summary

The TTID and TTFD are presented in the RUM Summary under Mobile Performance. The standalone Mobile Performance Dashboard also contains distribution visuals for TTID and TTFD. 

  {{< img src="real_user_monitoring/android/android-rum-summary-app-launch.png" alt="Android RUM Summary" style="width:90%;">}}

### Vital events

The time to initial display and time to full display are presented as vital events in the RUM session. They are also found under the first view after the `ApplicationLaunch` view. TTID and TTFD are captured if the user launches the application in the session. Neither the TTID nor the TTFD appear if the user does not launch the application during the session. 

  {{< img src="real_user_monitoring/android/android-app-launch-session.png" alt="Android session side panel" style="width:90%;">}}

The TTID and TTFD can be queried in the RUM Sessions Explorer using the following attributes on the vital event type:
- `@vital.type: app_launch`

> - `@vital.name: time_to_initial_display` or `@vital.name: time_to_full_display`

Reviewer 2: ```suggestion
- `@vital.type:app_launch`
- `@vital.name:time_to_initial_display` or `@vital.name:time_to_full_display`
```

since this code can be copy-pasted, it is better to remove spaces
